### PR TITLE
Allow to build QtWebKit as a static library

### PR DESCRIPTION
**Pull request by vitallium**

### CODE REVIEW

**Comment by annulen** [comment]
This patch somehow breaks loading resources, a number of tests fails

**Comment by annulen** [comment]
Most representative failing tests:
fast/dom/inner-text.html
fast/images/embed-image.html
fast/images/object-image.html
fast/images/pdf-as-image.html

File: Source/cmake/OptionsQt.cmake [hunk 1]
```diff
@@ -177,20 +186,21 @@ set(ENABLE_WEBKIT ON)
...
 endif ()
 
 if (SHARED_CORE)
     set(WebCoreTestSupport_LIBRARY_TYPE SHARED)
+    set(WebKit_LIBRARY_TYPE SHARED)
+    set(WebKitWidgets_LIBRARY_TYPE SHARED)
 else ()
     set(JavaScriptCore_LIBRARY_TYPE STATIC)
     set(WebCoreTestSupport_LIBRARY_TYPE STATIC)
+    set(WebKit_LIBRARY_TYPE STATIC)
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: This is wrong, WebKit_LIBRARY_TYPE is static only for QT_STATIC_BUILD. Also, it should be defined in WebKit/PlatformQt.cmake

File: Source/cmake/OptionsQt.cmake [hunk 2]
```diff
@@ -291,19 +306,49 @@ if (ENABLE_TEST_SUPPORT)
...
         macro_process_qtbase_prl_file(Qt5::${qt_module})
+        list(APPEND STATIC_LIB_DEPENDENCIES
+            ${Qt5${qt_module}_STATIC_LIB_DEPENDENCIES}
+        )
     endif ()
 endforeach ()
 
+if (QT_STATIC_BUILD)
+    list(REMOVE_DUPLICATES STATIC_LIB_DEPENDENCIES)
+    set(DEPEND_STATIC_LIBS ${STATIC_LIB_DEPENDENCIES})
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: Why do we need separate variables STATIC_LIB_DEPENDENCIES and DEPEND_STATIC_LIBS?

File: Source/cmake/OptionsQt.cmake [hunk 3]
```diff
@@ -340,10 +370,12 @@ set(CMAKE_AUTOMOC ON)
 
 # From OptionsEfl.cmake
 # Optimize binary size for release builds by removing dead sections on unix/gcc.
-if (COMPILER_IS_GCC_OR_CLANG AND UNIX AND NOT APPLE)
-    set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -ffunction-sections -fdata-sections")
-    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -ffunction-sections -fdata-sections -fno-rtti")
-    set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_SHARED_LINKER_FLAGS_RELEASE} -Wl,--gc-sections")
+if (COMPILER_IS_GCC_OR_CLANG AND UNIX)
+    if (NOT APPLE)
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: Looks like unrelated change

File: Source/cmake/OptionsQt.cmake [hunk 4]
```diff
@@ -291,19 +306,49 @@ if (ENABLE_TEST_SUPPORT)
...
+    link_directories("${_qt5Widgets_install_prefix}/lib")
 endif ()
+
 foreach (qt_module ${QT_OPTIONAL_COMPONENTS})
-    find_package("Qt5${qt_module}" ${REQUIRED_QT_VERSION})
+    find_package("Qt5${qt_module}" ${REQUIRED_QT_VERSION} PATHS "${Qt5_DIR}/.." NO_DEFAULT_PATH)
     if (QT_STATIC_BUILD)
         macro_process_qtbase_prl_file(Qt5::${qt_module})
+        list(APPEND STATIC_LIB_DEPENDENCIES
+            ${Qt5${qt_module}_STATIC_LIB_DEPENDENCIES}
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: Maybe just add those dependencies to `Qt5${qt_module}_LIBRARIES`?

File: Source/cmake/OptionsQt.cmake [hunk 5]
```diff
@@ -291,19 +306,49 @@ if (ENABLE_TEST_SUPPORT)
...
 if (QT_STATIC_BUILD)
     foreach (qt_module ${QT_REQUIRED_COMPONENTS})
         macro_process_qtbase_prl_file(Qt5::${qt_module})
+        list(APPEND STATIC_LIB_DEPENDENCIES
+            ${Qt5${qt_module}_STATIC_LIB_DEPENDENCIES}
+        )
     endforeach ()
+    # HACK: We must explicitly add LIB path of the Qt installation
+    # to correctly find qtpcre
+    link_directories("${_qt5Widgets_install_prefix}/lib")
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: _qt5_install_prefix

File: Source/cmake/OptionsQt.cmake [hunk 6]
```diff
@@ -394,9 +426,9 @@ if (NOT ENABLE_VIDEO)
 endif ()
 
 if (USE_QT_MULTIMEDIA)
-    find_package(Qt5Multimedia ${REQUIRED_QT_VERSION} REQUIRED)
+    find_package(Qt5Multimedia ${REQUIRED_QT_VERSION} REQUIRED PATHS "${Qt5_DIR}/.." NO_DEFAULT_PATH)
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: _qt5_install_prefix

File: Source/cmake/OptionsQt.cmake [hunk 7]
```diff
@@ -394,9 +426,9 @@ if (NOT ENABLE_VIDEO)
...
     # FIXME: Allow building w/o widgets
-    find_package(Qt5MultimediaWidgets ${REQUIRED_QT_VERSION} REQUIRED)
+    find_package(Qt5MultimediaWidgets ${REQUIRED_QT_VERSION} REQUIRED PATHS "${Qt5_DIR}/.." NO_DEFAULT_PATH)
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: Ditto

**Comment by annulen** [comment]
I've merged the main part of these changes, remainder is here

**Comment by Vitallium** [comment]
Ok. I'll rebase my branch and apply needed fixes.

**Comment by annulen** [comment]
Here is a rebased remainder after all conflicts solved, trust me ;)
